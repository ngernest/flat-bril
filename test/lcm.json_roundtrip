{
  "functions": [
    {
      "args": [
        {
          "name": "x",
          "type": "int"
        },
        {
          "name": "y",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "y"
          ],
          "dest": "greater",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "x",
            "y"
          ],
          "dest": "v4",
          "funcs": [],
          "labels": [],
          "op": "gt",
          "type": "bool"
        },
        {
          "args": [
            "v4"
          ],
          "funcs": [],
          "labels": [
            "then.1",
            "else.1"
          ],
          "op": "br"
        },
        {
          "label": "then.1"
        },
        {
          "args": [
            "x"
          ],
          "dest": "greater",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "endif.1"
          ],
          "op": "jmp"
        },
        {
          "label": "else.1"
        },
        {
          "label": "endif.1"
        },
        {
          "label": "foreverloop"
        },
        {
          "args": [
            "greater",
            "x"
          ],
          "dest": "modX",
          "funcs": [
            "getMod"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "args": [
            "greater",
            "y"
          ],
          "dest": "modY",
          "funcs": [
            "getMod"
          ],
          "labels": [],
          "op": "call",
          "type": "int"
        },
        {
          "dest": "zero",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "args": [
            "modX",
            "zero"
          ],
          "dest": "xZero",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "modY",
            "zero"
          ],
          "dest": "yZero",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "xZero",
            "yZero"
          ],
          "dest": "bothZero",
          "funcs": [],
          "labels": [],
          "op": "and",
          "type": "bool"
        },
        {
          "args": [
            "bothZero"
          ],
          "funcs": [],
          "labels": [
            "then.2",
            "else.2"
          ],
          "op": "br"
        },
        {
          "label": "then.2"
        },
        {
          "args": [
            "greater"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "loopend"
          ],
          "op": "jmp"
        },
        {
          "label": "else.2"
        },
        {
          "dest": "one",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "args": [
            "greater",
            "one"
          ],
          "dest": "greater",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "foreverloop"
          ],
          "op": "jmp"
        },
        {
          "label": "loopend"
        }
      ],
      "name": "main"
    },
    {
      "args": [
        {
          "name": "val",
          "type": "int"
        },
        {
          "name": "mod",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "val",
            "mod"
          ],
          "dest": "divisor",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "divisor",
            "mod"
          ],
          "dest": "multiple",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "val",
            "multiple"
          ],
          "dest": "rem",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "rem"
          ],
          "funcs": [],
          "labels": [],
          "op": "ret"
        }
      ],
      "name": "getMod",
      "type": "int"
    }
  ]
}
