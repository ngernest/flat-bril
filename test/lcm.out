@main
	op: id   		dest: greater		type: int		args: ["y"]	
	op: gt   		dest: v4		type: bool		args: ["x", "y"]	
	op: br   		args: ["v4"]		labels: ["then.1", "else.1"]
.then.1
	op: id   		dest: greater		type: int		args: ["x"]	
	op: jmp  		labels: ["endif.1"]
.else.1
.endif.1
.foreverloop
	op: call 		dest: modX		type: int		args: ["greater", "x"]		funcs: ["getMod"]
	op: call 		dest: modY		type: int		args: ["greater", "y"]		funcs: ["getMod"]
	op: const		dest: zero		type: int		value: 0	
	op: eq   		dest: xZero		type: bool		args: ["modX", "zero"]	
	op: eq   		dest: yZero		type: bool		args: ["modY", "zero"]	
	op: and  		dest: bothZero		type: bool		args: ["xZero", "yZero"]	
	op: br   		args: ["bothZero"]		labels: ["then.2", "else.2"]
.then.2
	op: print		args: ["greater"]	
	op: jmp  		labels: ["loopend"]
.else.2
	op: const		dest: one		type: int		value: 1	
	op: add  		dest: greater		type: int		args: ["greater", "one"]	
	op: jmp  		labels: ["foreverloop"]
.loopend
@getMod
	op: div  		dest: divisor		type: int		args: ["val", "mod"]	
	op: mul  		dest: multiple		type: int		args: ["divisor", "mod"]	
	op: sub  		dest: rem		type: int		args: ["val", "multiple"]	
	op: ret  		args: ["rem"]	
